language/move-core/types/src/resolver.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    account_address::AccountAddress,
+    language_storage::{ModuleId, StructTag},
+};
+use std::fmt::Debug;
+
+/// Traits for resolving Move modules and resources from persistent storage
+
+/// A persistent storage backend that can resolve modules by address + name.
+/// Storage backends should return
+///   - Ok(Some(..)) if the data exists
+///   - Ok(None)     if the data does not exist
+///   - Err(..)      only when something really wrong happens, for example
+///                    - invariants are broken and observable from the storage side
+///                      (this is not currently possible as ModuleId and StructTag
+///                       are always structurally valid)
+///                    - storage encounters internal error
+pub trait ModuleResolver {
+    type Error: Debug;
+
+    fn get_module(&self, id: &ModuleId) -> Result<Option<Vec<u8>>, Self::Error>;
+}
+
+/// A persistent storage backend that can resolve resources by address + type
+/// Storage backends should return
+///   - Ok(Some(..)) if the data exists
+///   - Ok(None)     if the data does not exist
+///   - Err(..)      only when something really wrong happens, for example
+///                    - invariants are broken and observable from the storage side
+///                      (this is not currently possible as ModuleId and StructTag
+///                       are always structurally valid)
+///                    - storage encounters internal error
+pub trait ResourceResolver {
+    type Error: Debug;
+
+    fn get_resource(
+        &self,
+        address: &AccountAddress,
+        typ: &StructTag,
+    ) -> Result<Option<Vec<u8>>, Self::Error>;
+}
```

Comment:
In general I'm fine with moving this into move-core-types, but I wonder if it's necessary to split the trait? The abstract Move global state contains both code and data, and I feel like we'll always implement both for any storage backends.

Reply:
Yes, good question. I think implementing both will indeed be a case. But some code I'm working on for inferring `MoveTypeLayout`'s from `StructTag`'s only needs a `ModuleResolver`, and I'd imagine that some static analysis tools will want to grab modules but not state.

In addition, could imagine client code that wants to decode resources into known schemas but doesn't care about modules.